src/ucp/core/ucp_mm.c

```diff
@@ -423,27 +447,26 @@ ucp_memh_get_slow(ucp_context_h context, void *address, size_t length,
 
     UCP_THREAD_CS_ENTER(&context->mt_lock);
     if (context->rcache == NULL) {
-        status = ucp_memh_get_nocache(context, address, length, &memh);
-        if (status != UCS_OK) {
-            goto out;
-        }
+        status = ucp_memh_create_no_rcache(context, reg_address, reg_length,
+                                           mem_type, &memh);
     } else {
-        status = ucs_rcache_get_unsafe(context->rcache, reg_address, reg_length,
-                                       PROT_READ | PROT_WRITE, NULL, &rregion);
-        if (status != UCS_OK) {
-            goto out;
-        }
+        status = ucp_memh_create_rcache(context, reg_address, reg_length,
+                                        mem_type, &memh);
+    }
 
-        memh        = ucs_derived_of(rregion, ucp_mem_t);
-        reg_address = (void*)rregion->super.start;
-        reg_length  = rregion->super.end - rregion->super.start;
+    if (status != UCS_OK) {
+        goto out;
     }
 
-    memh->mem_type = mem_type;
-    status         = ucp_memh_register(context, memh,
-                                       ~memh->md_map & reg_md_map,
-                                       reg_address, reg_length,
-                                       uct_flags);
+    /* Reinitialize address and length, because they could be greater in case
+     * a region which includes the searched region was found in the rcache */
+    reg_address = ucp_memh_address(memh);
+    reg_length  = ucp_memh_length(memh);
+
+    ucs_assert(memh->mem_type == mem_type);
```

Comment:
sorry, I guess it is not needed anymore, since we pass mem_type through rcache attributes